### PR TITLE
Add ClusterClaim API and include claims in cluster status

### PR DESCRIPTION
**Pull request by elgnay**
Link of enhancement proposal: https://github.com/open-cluster-management/enhancements/pull/2
Link of task: https://github.com/open-cluster-management/backlog/issues/7155

### CODE REVIEW

File: cluster/v1alpha1/types.go [hunk 1]
```diff
@@ -104,3 +104,40 @@ type ManagedClusterSetBindingList struct {
...
+	Spec ClusterClaimSpec `json:"spec,omitempty"`
+}
+
+type ClusterClaimSpec struct {
+	// Value is a claim-dependent string
+	Value string `json:"value,omitempty"`
+
+	// Signature may be used to validate a specific claim
+	// +optional
+	Signature string `json:"signature,omitempty"`
```
Comment (qiujian16) [inline cluster/v1alpha1/types.go]: remove the signature?

File: cluster/v1alpha1/types.go [hunk 2]
```diff
@@ -104,3 +104,40 @@ type ManagedClusterSetBindingList struct {
 	// Items is a list of ManagedClusterSetBinding.
 	Items []ManagedClusterSetBinding `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope="Cluster"
+
+// ClusterClaim represents cluster information that a managed cluster claims
```
Comment (qiujian16) [inline cluster/v1alpha1/types.go:113]: It is better that we also doc the wellknown name here I think

File: cluster/v1/types.go [hunk 3]
```diff
@@ -105,6 +109,15 @@ type ManagedClusterVersion struct {
 	Kubernetes string `json:"kubernetes,omitempty"`
 }
 
+// ManagedClusterClaim represents a ClusterClaim collected from managed cluster.
+type ManagedClusterClaim struct {
+	// Name is a well known or custom name to identify the claim.
```
Comment (qiujian16) [inline cluster/v1/types.go]: customized name?

File: cluster/v1/types.go [hunk 4]
```diff
@@ -105,6 +109,15 @@ type ManagedClusterVersion struct {
 	Kubernetes string `json:"kubernetes,omitempty"`
 }
 
+// ManagedClusterClaim represents a ClusterClaim collected from managed cluster.
```
Comment (qiujian16) [inline cluster/v1/types.go]: from a managed cluster

File: cluster/v1alpha1/types.go (hunk 1 above)
Comment (elgnay) [reply cluster/v1alpha1/types.go]: Removed.

File: cluster/v1alpha1/types.go (hunk 2 above)
Comment (elgnay) [reply cluster/v1alpha1/types.go:113]: Added.

File: cluster/v1/types.go (hunk 3 above)
Comment (elgnay) [reply cluster/v1/types.go]: Doc updated.

File: cluster/v1/types.go (hunk 4 above)
Comment (elgnay) [reply cluster/v1/types.go]: Doc updated.

File: cluster/v1/types.go [hunk 5]
```diff
@@ -95,6 +95,10 @@ type ManagedClusterStatus struct {
 
 	// Version represents the kubernetes version of the managed cluster.
 	Version ManagedClusterVersion `json:"version,omitempty"`
+
+	// ClusterClaims represents cluster information that a managed cluster claims
```
Comment (deads2k) [inline cluster/v1/types.go]: Indicate that this is written from the managed cluster (I assume it is) and provide a couple known examples please. 

Be sure to indicate that the set of claims is not uniform across a fleet, some claims can be vendor or version specific and may not be included in from all managed clusters.

File: cluster/v1/types.go [hunk 6]
```diff
@@ -105,6 +109,15 @@ type ManagedClusterVersion struct {
 	Kubernetes string `json:"kubernetes,omitempty"`
 }
 
+// ManagedClusterClaim represents a ClusterClaim collected from a managed cluster.
+type ManagedClusterClaim struct {
+	// Name is a well known or customized name to identify the claim.
```
Comment (deads2k) [inline cluster/v1/types.go]: give an example of the format of the name.  Also, if there is a regex for this and value in the upstream API, specify it here too.

File: cluster/v1alpha1/types.go [hunk 7]
```diff
@@ -104,3 +104,54 @@ type ManagedClusterSetBindingList struct {
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the attributes of the ClusterClaim.
+	Spec ClusterClaimSpec `json:"spec,omitempty"`
+}
+
+type ClusterClaimSpec struct {
+	// Value is a claim-dependent string
+	Value string `json:"value,omitempty"`
```
Comment (deads2k) [inline cluster/v1alpha1/types.go:130]: you should have a length restriction at the very least.

**Comment by deads2k** [comment]
fairly minor comments, but the validation is important, don't forget it.  You should also get the enhancement merged first.

/approve

File: cluster/v1/types.go (hunk 5 above)
Comment (elgnay) [reply cluster/v1/types.go]: Doc updated.

File: cluster/v1/types.go (hunk 6 above)
Comment (elgnay) [reply cluster/v1/types.go]: Examples and length restriction added.

File: cluster/v1alpha1/types.go (hunk 7 above)
Comment (elgnay) [reply cluster/v1alpha1/types.go:130]: Length restriction added.

File: cluster/v1/types.go [hunk 8]
```diff
@@ -95,6 +95,14 @@ type ManagedClusterStatus struct {
 
 	// Version represents the kubernetes version of the managed cluster.
 	Version ManagedClusterVersion `json:"version,omitempty"`
+
+	// ClusterClaims represents cluster information that a managed cluster claims,
+	// for example a unique cluster identifier (id.k8s.io) and kubernetes version
+	// (kubeversion.open-cluster-management.io). They are written from the managed
+	// cluster. The set of claims is not uniform across a fleet, some claims can be
+	// vendor or version specific and may not be included in from all managed clusters.
```
Comment (pmorie) [inline cluster/v1/types.go]: typo: "in from", should be "from"

**Review by pmorie — COMMENTED** [review]
one tiny nit.

File: cluster/v1/types.go (hunk 8 above)
Comment (elgnay) [reply cluster/v1/types.go]: The typo is corrected.

**Pull request merged**